Fix mintimeuuid() call that could crash Scylla

nyh opened the pull request:
This PR fixes the bug of certain calls to the `mintimeuuid()` CQL function which large negative timestamps could crash Scylla. It turns out we already had protections in place against very positive timestamps, but very negative timestamps could still cause bugs.

The actual fix in this series is just a few lines, but the bigger effort was improving the test coverage in this area. I added tests for the "date" type (the original reproducer for this bug used totimestamp() which takes a date parameter), and also reproducers for this bug directly, without totimestamp() function, and one with that function.

Finally this PR also replaces the assert() which made this molehill-of-a-bug into a mountain, by a throw.

Fixes #17035

ptrsmrn commented on test/cql-pytest/test_native_functions.py:
```diff
@@ -72,3 +72,49 @@ def test_blobas_wrong_size(cql, table1):
+# are 64-bit *signed* integers representing milliseconds since the epoch.
+# In particular, negative timestamps are allowed and should not be rejected.
+# The function tounixtimestamp can be used to extract the timestamp back from
+# the timeuuid. In this test we use "reasonably"-low timestamps to avoid
```
Would be good to give some hint as to what _reasonably-low_ means in this context, e.g. by referring to `timeuuid`'s precision.

ptrsmrn commented on test/cql-pytest/test_native_functions.py:
```diff
@@ -72,3 +72,49 @@ def test_blobas_wrong_size(cql, table1):
+# The function tounixtimestamp can be used to extract the timestamp back from
+# the timeuuid. In this test we use "reasonably"-low timestamps to avoid
+# running into problems with extreme values which we'll test in the next test.
+@pytest.mark.parametrize("timestamp", [-1706638779000, -123, 123, 1706638779000])
```
nitpick: could be parameterized by 2 values and 2 multipliers: a positive 1 and negative -1.

nyh replied:
I prefer not to. I want the tests I write to be readable easily, I don't want the reader to have to do these multiplications to understand what is really being tested. I guess it would make sense to do that if there were a lot of different combinations, but if there are only 4, I think it's clearer to just list them.

ptrsmrn commented on test/cql-pytest/test_type_date.py:
```diff
@@ -0,0 +1,90 @@
+    stmt = cql.prepare(f"INSERT INTO {table1} (p, d) VALUES (?, ?)")
+    cql.execute(stmt, [p, 123])
+    ds = list(cql.execute(f"SELECT d from {table1} where p = {p}"))
+    assert len(ds)==1
```
Don't we have some PEP8 check that'd eliminate such "glued" expressions? (In this case, by "glued" I mean that there's no whitespace between `==`).

ptrsmrn reviewed (approved):
Approved, FWIW

nyh commented:
Pushed a new version addressing review comments (whitespace and comment changes).

@bhalevy @kostja the second patch includes one low-level change (the main fix in this PR), introducing UUID_UNIXTIME_MIN, which is the area I believe both of you were interested in the past, so maybe you want to take a look.

denesb commented on utils/UUID_gen.hh:
```diff
@@ -383,7 +383,11 @@ public:
+        if ((0xf000000000000000UL & msb)) {
+            // We hope callers would try to avoid this case, but they don't
+            // always do, so assert() would be bad here - and caused #17035.
+            throw std::runtime_error("timeuuid time must fit in 60 bits");
```
Let's also include the bad input, so when we see it in the logs, we don't have to guess.

nyh replied:
Ok, I'll do that.

I didn't do this because 1. it was a pre-existing message, and 2. it's a header file and doesn't yet use fmt :-(
But I guess every source file in Scylla already uses fmt so adding it here won't hurt issue #1 because it won't really be included to any source file that doesn't already include it..

nyh replied:
Oh second thought, I'm not sure what it will help to print the input. It's usually already converted from some other format by the caller, and the problem in the after-conversion input is obvious - it has high-order bits on, we don't need to print it to see it. Or maybe you think seeing "msb" in full will be helpful?

What I think might be more useful is a backtrace, exactly what on_internal_error() does.
Yesterday I actually wanted to use on_internal_error here when I wrote this patch, but couldn't because it's a header file and I don't have a logger. This is why I used throw :-( Today I saw you also needed on_internal_error in a header file, so I think it's time to solve this problem. What do you say about me adding a utils::on_internal_error which has its own logger, and I'll use it here in the header file?

denesb replied:
I'm onboard. I think in general we should replace `assert()` with `on_internal_error()`.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6240/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6240/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7543/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6240/)
#### Failed Tests (3/24297):
- [test_remove_garbage_members_from_group0_after_abort_removenode[made_node.*a_non-voter_in_group_0-]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6240/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_remove_garbage_members_from_group0_after_abort_removenode_made_node__a_non_voter_in_group_0__) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_members_from_group0_after_abort_removenode[made_node.*a_non-voter_in_group_0-])
- [test_cannot_disable_cluster_feature_after_all_declare_support](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6240/testReport/junit/%28root%29/test_raft_cluster_features/Tests___Unit_Tests___test_cannot_disable_cluster_feature_after_all_declare_support) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cannot_disable_cluster_feature_after_all_declare_support)
- [topology_experimental_raft.test_raft_cluster_features.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6240/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_cluster_features_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_cluster_features.debug.2)
#### Build Details:
- Duration: 3 hr 18 min
- Builder: i-0658bed00a9548b58 (m5d.8xlarge)

nyh commented:
Pushed a new version which introduces a new utils::on_external_error() and uses it instead of an exception. Botond, the exception now includes a full backtrace, but still not the problematic timestamp (which I thought is not useful) - please let me know if you still think I should add it.

denesb commented:
> Pushed a new version which introduces a new utils::on_external_error() and uses it instead of an exception. Botond, the exception now includes a full backtrace, but still not the problematic timestamp (which I thought is not useful) - please let me know if you still think I should add it.

I think the backtrace is more useful. Indeed `msb` doesn't look very useful in determining what was wrong, we can already guess this from the check.
In other words, LGTM.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6251/)
#### Build Details:
- Duration: 15 min
- Builder: spider8.cloudius-systems.com

nyh commented:
Link of of build/debug/test/boost/UUID_test failed:
```
[2024-01-31T13:46:17.868Z] ld.lld: error: undefined symbol: utils::on_internal_error(std::basic_string_view<char, std::char_traits<char>>)
```
It's because this new on_internal_error function comes from utils/on_internal_error, which I guess wasn't linked with the test. I'll send a fix.

nyh commented:
Pushed a new version fixing the build of UUID_test.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6255/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6255/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7555/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6255/)
#### Build Details:
- Duration: 3 hr 18 min
- Builder: i-0fbe550acaf266941 (m5d.8xlarge)

mykaul commented:
@ptrsmrn - please review as well.

ptrsmrn commented:
> @ptrsmrn - please review as well.

@mykaul I already approved.
